### ISSUE TITLE
Ignite is missing a cast message

### ISSUE THREAD

**Issue by rpherbig** [rpherbig/dr-scripts#504]
From Qetu:
```
[combat-trainer]>cast my spear
The flames dancing along your fingertips flash with intensity before curling down along the length of your narrow-headed spear.
The complementary nature of the spell empowers you.
>
* Moving with great skill, a S'lai scout sidesteps and bashes at you.  You dodge, barely stepping to one side.  
[You're bruised, winded, adeptly balanced and in better position.]
>
* A S'lai scout thrusts a quarter staff at you.  You evade, barely leaning out of harm's reach.  
[You're bruised, winded, adeptly balanced and in good position.]
>
* Moving well, a S'lai scout lunges a quarter staff at you.  You evade, barely stepping to one side.  
[You're bruised, winded, adeptly balanced and in good position.]
>
The S'lai scout moves into a position to parry.
>
[hunting-buddy: ***STATUS*** 38 minutes of hunting, still waiting on Polearms]
>
A S'lai scout appears less distracted.
>
The S'lai scout moves into a position to parry.
>
A S'lai scout suddenly slips on the wet ground and falls down!
>
The S'lai scout moves into a position to parry.
>
[combat-trainer: *** No match was found after 15 seconds, dumping info]
[combat-trainer: messages seen length: 19]
[combat-trainer: message: The flames dancing along your fingertips flash with intensity before curling down along the length of your narrow-headed spear.]
[combat-trainer: message: The complementary nature of the spell empowers you.]
[combat-trainer: message: Substratum  (9 roisaen)]
[combat-trainer: message: Ignite  (4 roisaen)]
[combat-trainer: message: Ethereal Shield  (2 roisaen)]
[combat-trainer: message: * Moving with great skill, a S'lai scout sidesteps and bashes at you.  You dodge, barely stepping to one side.  ]
[combat-trainer: message: [You're bruised, winded, adeptly balanced and in better position.]]
[combat-trainer: message: * A S'lai scout thrusts a quarter staff at you.  You evade, barely leaning out of harm's reach.  ]
[combat-trainer: message: [You're bruised, winded, adeptly balanced and in good position.]]
[combat-trainer: message: Substratum  (9 roisaen)]
[combat-trainer: message: Ignite  (4 roisaen)]
[combat-trainer: message: Ethereal Shield  (2 roisaen)]
[combat-trainer: message: * Moving well, a S'lai scout lunges a quarter staff at you.  You evade, barely stepping to one side.  ]
[combat-trainer: message: [You're bruised, winded, adeptly balanced and in good position.]]
[combat-trainer: message: The S'lai scout moves into a position to parry.]
[combat-trainer: message: A S'lai scout appears less distracted.]
[combat-trainer: message: The S'lai scout moves into a position to parry.]
[combat-trainer: message: A S'lai scout suddenly slips on the wet ground and falls down!]
[combat-trainer: message: The S'lai scout moves into a position to parry.]
[combat-trainer: checked against [/^You gesture/i, /You reach with both/i, /You listen intently to the air/i, /You close your eyes/i, /You close yours eyes/i, /Your target pattern dissipates/i, /You make a holy gesture/i, /You raise your left hand skyward/i, /Flickering shadows tear away/i, /You sense a Malediction/i, /You release an accompaniment/i, /You roll your hands in an elliptical pattern/i, /You cup your hand/i, /You raise your right hand skyward/i, /You flicker your malevolent desires towards/i, /You clench your hands into fists and grit your teeth/i, /You make a sweeping gesture with your arm/i, /You wave your hand/i, /^Your target pattern dissipates/, /^You can't cast that at yourself/]]
[combat-trainer]>perc
You reach out with your developing senses and see muted (2/5) streams of roiling Elemental mana coursing through the area.
You sense the Substratum spell upon you, which will last for about eight roisaen.
You sense the Ignite spell upon you, which will last for about three roisaen.
You sense the Ethereal Shield spell upon you, which will last for about two roisaen.
Roundtime: 7 sec.
```

This is his `buff_spells` (http://pastebin.com/bN0K7X16):
```
 Ignite:
    abbrev: ignite
    recast: 0
    mana: 5
    cast: cast my spear
```

**Comment by rpherbig** [comment]
I think that the special Ignite logic is getting ignored and it's being cast as a normal buff. You can see that it uses the cast message `cast my spear` (from the buff info) instead of `cast checkright` (from the special ignite logic).

**Comment by rpherbig** [comment]
It seems like Ignite (and really, this issue is for Bless too) should be removed from @buff_spells. And I'm pondering if we can get rid of the special methods and just do a check when casting to use a custom cast message (not to be confused with the `cast` setting, which we can't use here because the weapon name changes at run-time). Maybe we need a new setting to indicate which spells can be snap casted?

**Comment by rpherbig** [comment]
Debugging and further discussion has revealed that the root cause is that Ignite is (currently) snap casted. In the happy path, `check_ignite` snap casts Ignite and `check_buffs` sees that it is an active spell and never attempts to cast it.

However, Qetu is low circle and the snap cast failed. Then `check_buffs` saw the buff missing and attempted to cast it normally.

I think the right approach here is to get rid of `check_ignite` and let Ignite go through the normal cast process. There are three problems with that: 1) Ignite needs a custom cast message determined at run-time; 2) Ignite must be released before it can be recast; and 3) Ignite cannot be cast on aimed weapons.

I will probably add a `snap_cast` setting to make this spell easier to use when hunting.

**Comment by rpherbig** [comment]
Integrating something like Ignite with our normal cast logic requires a few different integration points. I hope they're clear from this. While I don't like having something so specific as `game_state.casting_ignite`, this should be able to be generalized to the other weapon buffs (Resonance, Rutilor's Edge, etc.).

Possible next steps:
- Do not release/recast on summoned weapons (they keep Ignite even after shaping)
- Can we use `expire` instead of `recast: -1`?
- Add a `snap:` sub-setting for each spell (this functionality, specific to Ignite, was lost in my rewrite, but really should be generalized)
- Generalize this logic to support the other weapon buffs

I'm going to leave this issue open until 1) some people in-game can confirm it is working, and 2) @rcuhljr has had a chance to review this.

**Comment by rcuhljr** [comment]
Why is the solution just don't use ignite until you can snap cast it? It's kind of a bad option with our exp/hit thresholds given that casting waits on full prep.

**Comment by rpherbig** [comment]
We used to require snap casting. Now we do not.

Snap casting should still be an option for those that want it.

See #510

**Comment by rcuhljr** [comment]
Sorry that was a typo, I was trying to ask why /isn't/. But the works already done it seems so not worth worrying about.

**Comment by rpherbig** [comment]
Gotcha. Two users noted to me that it didn't work unless you could snap cast, and they both used 1 or 2 weapons in total, which means Ignite is not recast very often.
